go/libraries/doltcore/doltdb/commit_hooks.go

```diff
@@ -93,6 +94,97 @@ func pushDataset(ctx context.Context, destDB, srcDB datas.Database, tempTableDir
 	return err
 }
 
+type PushArg struct {
+	ds   datas.Dataset
+	db   datas.Database
+	hash hash.Hash
+}
+
+type AsyncPushOnWriteHook struct {
+	out io.Writer
+	ch  chan PushArg
+}
+
+const asyncPushBufferSize = 100
+
+var _ datas.CommitHook = (*AsyncPushOnWriteHook)(nil)
+
+// NewAsyncPushOnWriteHook creates a AsyncReplicateHook
+func NewAsyncPushOnWriteHook(ctx context.Context, destDB *DoltDB, tmpDir string) *AsyncPushOnWriteHook {
+	ch := make(chan PushArg, asyncPushBufferSize)
+
+	var newHeads = make(map[string]PushArg, asyncPushBufferSize)
+	go func() error {
+		defer close(ch)
+		for {
+			p, ok := <-ch
+			if !ok {
+				return ctx.Err()
+			}
+			newHeads[p.ds.ID()] = p
+		}
+	}()
+
+	go func() error {
+		var latestHeads = make(map[string]hash.Hash, asyncPushBufferSize)
+		for {
+			select {
+			case <-ctx.Done():
+				return ctx.Err()
+			default:
+				if len(newHeads) == 0 {
+					continue
+				}
+				// TODO do i need to explicitly copy newHeads?
+				for id, newCm := range newHeads {
+					if latest, ok := latestHeads[id]; !ok || latest != newCm.hash {
+						err := pushDataset(ctx, destDB.db, newCm.db, tmpDir, newCm.ds)
+						if err != nil {
+							// TODO what to do with errors?
+							return err
+						}
+						latestHeads[id] = newCm.hash
+					}
+				}
+			}
+		}
+	}()
+
+	return &AsyncPushOnWriteHook{ch: ch}
+}
+
+// Execute implements datas.CommitHook, replicates head updates to the destDb field
+func (ah *AsyncPushOnWriteHook) Execute(ctx context.Context, ds datas.Dataset, db datas.Database) error {
+	rf, ok, err := ds.MaybeHeadRef()
+	if !ok {
```

Comment:
should handle errors before other return parameters as the other return parameters are often times not valid when an error is returned.  Doesn't really matter in this case, but still a better practice.